src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java

```diff
@@ -335,6 +338,41 @@ private ComponentMetadata getComponentMetadata(ComponentIdentifier componentIden
         });
     }
 
+    /**
+     * Check if all plugins that are required to execute pre-merge steps like download for
+     * other components are included in the deployment.
+     *
+     * @param pkgIds deployment dependency closure
+     * @throws MissingRequiredComponentsException when any required plugins are not included
+     * @throws PackageLoadingException when other errors occur
+     */
+    public void checkPreparePackagesPrerequisites(List<ComponentIdentifier> pkgIds)
+            throws MissingRequiredComponentsException, PackageLoadingException {
+        for (ComponentIdentifier pkgId : pkgIds) {
+            Optional<ComponentRecipe> recipeOption = componentStore.findPackageRecipe(pkgId);
+            if (!recipeOption.isPresent()) {
+                throw new PackageLoadingException(
+                        "Unexpected error - cannot find recipe for a component to be " + "prepared");
+            }
+            for (ComponentArtifact artifact : recipeOption.get().getArtifacts()) {
+                // TODO : Make this more generic & use dedicated component type
+                if (artifact.getArtifactUri().getSchemeSpecificPart().equals(ArtifactDownloaderFactory.DOCKER_SCHEME)) {
+                    // TODO : Find better name for component
+                    if (!pkgIds.contains("aws.greengrass.DockerPlugin")) {
```

Comment:
1) Fixed it.

2) The plugin will be required at download time, hence checking the dependency closure since even if the component is not a dependency of other components, it can still be present in the deployment independently

3) The plugin code is currently in the nucleus like in this PR, but ultimately, it needs to be moved out of the nucleus like other plugins. And for that, we need a check for it always having to be included in deployment for backwards compatibility when we move it out

Reply:
Does it mean customer need at least two deployments to use this feature? first deployment to add docker plugin, and second to deploy a docker artifact component.
If that's the case, instead of finding `aws.greengrass.DockerPlugin` in the resolved component list, we should find in `kernel.locate` in order to make sure it's already loaded and running